File: DynamicProgramming/53_MaximumSubarray.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-07 11:08:44
 */


class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int nums_len = nums.size();
        int pre_sum = nums[0];
        int max_sum = nums[0];

        for(int i=1;i<nums_len;i++){
            pre_sum = max(pre_sum+nums[i], nums[i]);
            max_sum = max(max_sum, pre_sum);
        }
        return max_sum;
    }
};

/*
[-1]
[1]
[-9,-2,-3,-5,-3]
[-2,1,-3,4,-1,2,1,-5,4]
*/

File: DynamicProgramming/53_MaximumSubarray.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


class Solution(object):
    # O(n) space
    def maxSubArray(self, nums):
        num_len = len(nums)
        # dp[i]: Largest sum of contiguous subarray start from i
        dp = [-1] * num_len
        max_sum = dp[num_len - 1] = nums[num_len - 1]

        for i in range(num_len - 2, -1, -1):
            dp[i] = max(nums[i], dp[i+1]+nums[i])
            max_sum = max(dp[i], max_sum)

        return max_sum


class Solution_2(object):
    # DP same with the previous, but O(1) space
    def maxSubArray(self, nums):
        num_len = len(nums)
        max_sum = pre_sum = nums[num_len - 1]

        for i in range(num_len - 2, -1, -1):
            pre_sum = max(nums[i], pre_sum+nums[i])
            max_sum = max(pre_sum, max_sum)

        return max_sum

"""
[-1]
[1]
[-9,-2,-3,-5,-3]
[-2,1,-3,4,-1,2,1,-5,4]
"""
